src/main/kotlin/com/pubnub/api/crypto/cryptor/HeaderParser.kt

```diff
@@ -0,0 +1,127 @@
+private const val SENTINEL_ENDING_INDEX = 3
+private const val VERSION_INDEX = 4
+private const val CRYPTOR_ID_STARTING_INDEX = 5
+private const val CRYPTOR_ID_ENDING_INDEX = 8
+private const val CRYPTOR_DATA_SIZE_STARTING_INDEX = 9
+private const val THREE_BYTES_CRYPTOR_DATA_SIZE_STARTING_INDEX = 10
+private const val THREE_BYTES_CRYPTOR_DATA_SIZE_ENDING_INDEX = 11
+private const val MAX_VALUE_THAT_CAN_BE_STORED_ON_TWO_BYTES = 65535
+
+class HeaderParser {
+    private val log = LoggerFactory.getLogger(HeaderParser::class.java)
+
+    fun parseHeader(data: ByteArray): ParseResult {
+
+        val sentinel = data.sliceArray(SENTINEL_STARTING_INDEX..SENTINEL_ENDING_INDEX)
+        if (!SENTINEL.contentEquals(sentinel)) {
+            return ParseResult.NoHeader
+        }
+
+        if (data.size < MINIMAL_SIZE_OF_DATA_HAVING_CRYPTOR_HEADER) {
+            throw PubNubException(errorMessage = "Minimal size of Cryptor Data Header is: $MINIMAL_SIZE_OF_DATA_HAVING_CRYPTOR_HEADER")
+        }
+
+        validateCryptorHeaderVersion(data)
+
+        val cryptorId = data.sliceArray(CRYPTOR_ID_STARTING_INDEX..CRYPTOR_ID_ENDING_INDEX)
+        log.trace("CryptoId: ${String(cryptorId, Charsets.UTF_8)}")
+
+        val cryptorDataSizeFirstByte: Byte = data[CRYPTOR_DATA_SIZE_STARTING_INDEX]
+        val (startingIndexOfCryptorData, cryptorDataSize) = getCryptorDataSizeAndStartingIndex(
+            data,
+            cryptorDataSizeFirstByte
+        )
+
+        if (startingIndexOfCryptorData + cryptorDataSize > data.size) {
+            throw PubNubException(errorMessage = "Input data size: ${data.size} is to small to fit header of size $startingIndexOfCryptorData and cryptorData of size: $cryptorDataSize")
+        }
+        val cryptorData =
+            data.sliceArray(startingIndexOfCryptorData until (startingIndexOfCryptorData + cryptorDataSize))
+        val sizeOfCryptorHeader = startingIndexOfCryptorData + cryptorDataSize
+        val encryptedData = data.sliceArray(sizeOfCryptorHeader until data.size)
+
+        return ParseResult.Success(cryptorId, cryptorData, encryptedData)
+    }
+
+    fun createCryptorHeader(cryptorId: ByteArray, cryptorData: ByteArray?): ByteArray {
+        val sentinel: ByteArray = SENTINEL
+        val cryptorHeaderVersion: Byte = getCurrentCryptoHeaderVersion()
+        val cryptorDataSize: Int = cryptorData?.size ?: 0
+        val finalCryptorDataSize: ByteArray =
+            if (cryptorDataSize < THREE_BYTES_SIZE_CRYPTOR_DATA_INDICATOR.toInt()) {
+                byteArrayOf(cryptorDataSize.toByte()) // cryptorDataSize will be stored on 1 byte
+            } else if (cryptorDataSize < MAX_VALUE_THAT_CAN_BE_STORED_ON_TWO_BYTES) {
+                byteArrayOf(cryptorDataSize.toByte()) + writeNumberOnTwoBytes(cryptorDataSize) // cryptorDataSize will be stored on 3 byte
+            } else {
+                throw PubNubException(errorMessage = "Something is wrong with Cryptor Data Size: $cryptorDataSize")
```

Comment:
We could be more specific here. We know what's wrong with the size. It's too large

Reply:
Sure. We can go with:
throw PubNubException(errorMessage = "Cryptor Data Size is: $cryptorDataSize wheareas max cryptor data size is: $MAX_VALUE_THAT_CAN_BE_STORED_ON_TWO_BYTES ")